src/atc/main.go

```diff
@@ -0,0 +1,239 @@
+}
+
+func GetEnv(name, def string) string {
+	if value := os.Getenv(name); value == "" {
+		return def
+	} else {
+		return value
+	}
+}
+
+func main() {
+	thriftAddr := kingpin.Flag("thrift-addr", "thrift server address").Short('T').Default(GetEnv("ATCD_ADDR", "127.0.0.1:9090")).TCP()
+	thriftProto := kingpin.Flag("thrift-proto", "thrift server protocol").Short('P').Default(GetEnv("ATCD_PROTO", "json")).String()
+
+	var (
+		defMember = GetEnv("ATC_MEMBER", "")
+
+		// Duplicate flags/args
+		token  string
+		id     int64
+		member net.IP
+	)
+
+	kingpin.Command("info", "Prints info about the ATC shaping")
+
+	group := kingpin.Command("group", "Manage atc shaping groups")
+
+	group.Command("list", "List groups")
+
+	groupAdd := group.Command("create", "create a group")
+	if defMember != "" {
+		groupAdd.Arg("member", "IP address of the member").Default(defMember).IPVar(&member)
+	} else {
+		groupAdd.Arg("member", "IP address of the member").Required().IPVar(&member)
+	}
+
+	groupShow := group.Command("show", "show info about a group")
+	groupShow.Arg("id", "id of the group").Required().Int64Var(&id)
+
+	groupJoin := group.Command("join", "leave a group")
+	groupJoin.Flag("token", "token").Short('t').Default("").StringVar(&token)
+	groupJoin.Arg("id", "id of the group").Required().Int64Var(&id)
+	if defMember != "" {
+		groupJoin.Arg("member", "IP address of the member").Default(defMember).IPVar(&member)
+	} else {
+		groupJoin.Arg("member", "IP address of the member").Required().IPVar(&member)
+	}
+
+	groupLeave := group.Command("leave", "leave a group")
+	groupLeave.Flag("token", "token").Short('t').Default("").StringVar(&token)
+	groupLeave.Arg("id", "id of the group").Required().Int64Var(&id)
+	if defMember != "" {
+		groupLeave.Arg("member", "IP address of the member").Default(defMember).IPVar(&member)
+	} else {
+		groupLeave.Arg("member", "IP address of the member").Required().IPVar(&member)
+	}
+
+	groupToken := group.Command("token", "get a token")
+	groupToken.Arg("id", "id of the group").Required().Int64Var(&id)
+
+	groupUnshape := group.Command("unshape", "remove shaping from a group")
+	groupUnshape.Flag("token", "token").Short('t').Default("").StringVar(&token)
+	groupUnshape.Arg("id", "id of the group").Required().Int64Var(&id)
+
+	groupShape := group.Command("shape", "apply shaping to a group")
+	groupShape.Flag("token", "token").Short('t').Default("").StringVar(&token)
+
+	// Uplink shaping
+	upRate := groupShape.Flag("up.rate", "uplink rate in Kb/s").Default("0").Int32()
+	upDelay := groupShape.Flag("up.delay", "uplink delay in ms").Default("0").Int32()
+	upLoss := groupShape.Flag("up.loss", "uplink loss percentage").Default("0").Float64()
+
+	// Downlink shaping
+	dnRate := groupShape.Flag("dn.rate", "downlink rate in Kb/s").Default("0").Int32()
+	dnDelay := groupShape.Flag("dn.delay", "downlink delay in ms").Default("0").Int32()
+	dnLoss := groupShape.Flag("dn.loss", "downlink loss percentage").Default("0").Float64()
```

Comment:
Right now only support bandwidth, delay and package loss basics. Can easily add more fields later.